[CuTe] Add type for CuTe layout via claude

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #161106
* #161016
* #162690
* #162414
* __->__ #162534
* #162413

This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.
We removed all suppressions of linters and add type checker, type alias and mypy ignore(if needed) so that the pycute code will be checked by linter.


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

fduwjj committed:
[CuTe] Add type for CuTe layout via claude

[ghstack-poisoned]

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,20 +37,24 @@
+IntTuple = Union[int, tuple[Any, ...]]
+
+
+def is_int(x: Any) -> bool:
```
I won't repeat this for other Any instances

ezyang commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -206,18 +214,21 @@ def crd2crd(crd, dst_shape, src_shape=None):
 
 # Filter trg according to crd: keep only elements of trg that are paired with None
-def slice_(crd: Union[None, tuple, int], trg: Union[tuple, int]):
+def slice_(crd: Union[None, tuple, int], trg: Union[tuple, int]) -> Union[tuple, int]:
```
This type is kind of bad but the original type was also kind of bad so whatever

ezyang commented on torch/distributed/_pycute/layout.py:
```diff
@@ -312,7 +320,9 @@ def left_inverse(layout):
 # Use tuples-of-layouts to perform this operation by-mode and None as no-op
-def logical_divide(layoutA, layoutB):
+def logical_divide(
+    layoutA: Layout, layoutB: Optional[Union[Layout, IntTuple, tuple[Any, ...]]]
```
An alias for  Optional[Union[Layout, IntTuple, tuple[Any, ...]]] would be better; also the Any is too grabby

ezyang commented on torch/distributed/_pycute/layout.py:
```diff
@@ -385,28 +401,36 @@ def hier_unzip(splitter, layoutA, layoutB):
 
 
-def slice_and_offset(crd: tuple, layout: Layout):
+def slice_and_offset(crd: tuple[Any, ...], layout: Layout) -> tuple[Layout, int]:
```
Any here can be improved

ezyang commented:
You should be able to feed my CR comments directly to Claude. Please review the output yourself though.

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_pycute/layout.py:
```diff
@@ -38,39 +35,61 @@
+    flatten,
+    has_none,
+    IntTuple,
+    is_int,
```
Please explicitly type as TypeAlias. Removes ambiguity for type checker and IDEs.

Skylion007 commented on torch/distributed/_pycute/layout.py:
```diff
@@ -38,39 +35,61 @@
+)
+
+
+# Type aliases
```
TypeIs here too

fduwjj commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,20 +37,24 @@
     return isinstance(x, Integer)
 
 
-def is_tuple(x):
```
hmm we need to support python3.9 so we probably cannot use TypeIs here.

fduwjj commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,72 +36,76 @@
+IntTuple = Union[int, tuple["IntTuple", ...]]
+
+
+def is_int(x: object) -> TypeGuard[int]:
```
I see we have CI to test python3.9 so TypeIs might be available in 3.9?

Skylion007 replied:
It's available via the typing_extensions backport and already used extensively throughout the codebase.

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,72 +36,76 @@
 from functools import reduce
 from itertools import chain
-from typing import Union
+from typing import Optional, TypeAlias, TypeGuard, Union
```
right this works for python 3.10+ (or 3.11+)

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,72 +36,77 @@
+IntTuple: TypeAlias = Union[int, tuple["IntTuple", ...]]
+
+
+def is_int(x: object) -> TypeGuard[int]:
```
```suggestion
def is_int(x: object) -> TypeIs[int]:

fduwjj replied:
Hmm I think as I have explained, TypeIS is not available in python3.9 while we have CI for that. So we won't take your comment for now. And down the road, if we bump python version for PT, We will definitely use TypeIs. Does that sound good?

Skylion007 replied:
@fduwjj It is available through `typing_extensions`

Skylion007 commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,72 +36,77 @@
 from itertools import chain
-from typing import Union
+from typing import Optional, Union
+from typing_extensions import TypeAlias, TypeGuard
```
```suggestion
from typing_extensions import TypeAlias, TypeIs

Skylion007 commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,72 +36,77 @@
             return tuple(i for a in t for i in flatten(a))
     else:
-        return (t,)
+        return (t,)  # type: ignore[return-value]  # t is int, converted to tuple[int]
```
```suggestion
        return (t,)
```
Wont be necessary with TypeIs

Skylion007 commented on torch/distributed/_pycute/int_tuple.py:
```diff
@@ -39,72 +36,77 @@
 
 
-def is_tuple(x):
+def is_tuple(x: object) -> TypeGuard[tuple]:
```
```suggestion
def is_tuple(x: object) -> TypeIs[tuple]:

Skylion007 commented on torch/distributed/_pycute/layout.py:
```diff
@@ -38,39 +35,62 @@
-
-from .int_tuple import *
+from typing import Optional, Union
+from typing_extensions import TypeAlias, TypeGuard
```
```suggestion
from typing_extensions import TypeAlias, TypeIs
```

Skylion007 commented on torch/distributed/_pycute/layout.py:
```diff
@@ -38,39 +35,62 @@
 
 
-def is_layout(x):
+def is_layout(x: object) -> TypeGuard["Layout"]:
```
```suggestion
def is_layout(x: object) -> TypeIs["Layout"]:
```

Skylion007 reviewed (commented):
Use TypeIs from here from the typing_extensions, which is available on Python 3.9 as it's a pip package which is a dependency of PyTorch and is not part of the standard library: `typing` which is bundled in 3.9

ezyang reviewed (approved):
You should try to do Aaron's suggestions, but I will go ahead and preemptively approve here.

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.
We removed all suppressions of linters and add type checker, type alias and mypy ignore(if needed) so that the pycute code will be checked by linter.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.
We removed all suppressions of linters and add type checker, type alias and mypy ignore(if needed) so that the pycute code will be checked by linter.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[CuTe] Add type for CuTe layout via claude"


This PR mostly is a cosmetic change using Claude to add types for copied PyCute code.
We removed all suppressions of linters and add type checker, type alias and mypy ignore(if needed) so that the pycute code will be checked by linter.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]
